### PR TITLE
Median[], RankedMin[], RankedMax[]

### PR DESCRIPTION
**Pull request by poke1024**
This introduces linear time selection capabilities based on introselect (https://en.wikipedia.org/wiki/Introselect).

I implemented introselect from scratch, since the only two implementations I'm aware of (one in C++'s STL and one inside numpy's argpartition) seem quite cryptic to me in regards to certain steps, and - apart from licensing issues - a Python port of these would have been even more cryptic. For the most part, I just implemented Hoare's old algorithms as described in the 1960 ACM publication without any changes, and, indeed, they work.

Having a reliable median implementation is an important stepping stone for implementing something like kdtrees later on.

The new package "algorithm" is of course just a proposal: for functions implementing basic algorithms and data structures that are too bulky (and too reliant on private functions) to be included in the regular files.

### CODE REVIEW

File: mathics/algorithm/introselect.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        return True
+
+    def test_range(l_max):
+        # we test two cases: many same elements, and few same elements.
+        return test_configuration(l_max, 10) and test_configuration(200, 1000)
+
+    if test_range(200):
+        print('ALL OK.')
+    else:
+        print('ABORTED WITH FAILURE.')
```
Comment (sn6uv) [inline mathics/algorithm/introselect.py]: I think a better place for these tests would be e.g. `test/algorithm/test_introselect.py`.

**Comment by sn6uv** [comment]
Good stuff. Did some local testing and it is indeed faster for large lists:
```
In [11]: a = [random.randint(1, 10**5) for _ in range(10**6)]

In [12]: %timeit introselect(a, 500)
1 loop, best of 3: 366 ms per loop

In [13]: %timeit sorted(a)[500]
1 loop, best of 3: 630 ms per loop
```

File: mathics/builtin/lists.py [hunk 2]
```diff
@@ -2696,3 +2697,113 @@ def apply(self, expr, levels, evaluation):
...
+    >> Median[{-11, 38, 501, 1183}]
+     = 539 / 2
+
+    Passing a matrix returns the medians of the respective columns:
+    >> Median[{{100, 1, 10, 50}, {-1, 1, -2, 2}}]
+     = {99 / 2, 1, 4, 26}
+    """
+
+    messages = {
+        'rectn': 'Expected a list of numbers or a list of lists at position 1 in ``.'
```
Comment (sn6uv) [inline mathics/builtin/lists.py]: The mathematica `Medain::rectn` message is ```Rectangular array of real numbers is expected at position `2` in `1`.```, I think we should say something about requiring the input to be rectangular.

**Comment by poke1024** [comment]
The `%timeit` results are great to see. The differences in a Mathics context should be even bigger since comparing `BaseExpression`s is expensive (in contrast to comparing integers) and so the smaller number of comparisons (in contrast to `sort`) should have an even bigger effect.

**Comment by poke1024** [comment]
I've changed `_median5` to only do 6 comparisons; this is only used in a border case, but who knows...

**Comment by poke1024** [comment]
Yes, ready to merge as far as I'm concerned.

**Pull request merged**